lib/_http_incoming.js

```diff
@@ -114,19 +114,20 @@ IncomingMessage.prototype._read = function _read(n) {
     readStart(this.socket);
 };
 
-
 // It's possible that the socket will be destroyed, and removed from
 // any messages, before ever calling this.  In that case, just skip
 // it, since something else is destroying this connection anyway.
-IncomingMessage.prototype.destroy = function destroy(error) {
-  // TODO(ronag): Implement in terms of _destroy
-  this.destroyed = true;
-  if (this.socket)
-    this.socket.destroy(error);
-  return this;
+IncomingMessage.prototype._destroy = function _destroy(err, cb) {
+  if (!this.readableEnded || !this.complete) {
+    this.aborted = true;
+    this.emit('aborted');
+  }
+  if (this.socket && !this.readableEnded) {
+    this.socket.destroy(err);
+  }
+  this.listenerCount('error') > 0 ? cb(err) : cb();
```

Comment:
With this approach, we can emulate a backward-compatible behavior. To be fair though, this does not behave exactly like [before](https://github.com/nodejs/node/blob/master/lib/_http_server.js#L552) because this check was done only when terminating early, while now it is done every time `_destroy` is called with an error. It would be an easy change to move this check a few lines above, if necessary.

Reply:
Why not just do this inside !this.readableEnded?